src/Hosting/Abstractions/src/Microsoft.AspNetCore.Hosting.Abstractions.csproj

```diff
@@ -15,6 +16,8 @@
     <Reference Include="Microsoft.AspNetCore.Hosting.Server.Abstractions" />
     <Reference Include="Microsoft.AspNetCore.Http.Abstractions" />
     <Reference Include="Microsoft.Extensions.Hosting.Abstractions" />
+
+    <Compile Include="$(SharedSourceRoot)SkipLocalsInit.cs" LinkBase="Shared\SkipLocalsInit.cs" />
```

Comment:
Hosting.Abstractions is not on a perf-critical code path and from your review it doesn't contain any stackallocs. Why did you mark this assembly? The same goes for Server.Abstractions and likely some of the others.

Reply:
I added it to all projects referenced by Kestrel.Core. Want them removed?

Reply:
What does it get us if we're not using stackallocs in these assemblies?

Reply:
@stephentoub Is there much benefit to SkipLocalsInit outside of stackalloc usages?

Reply:
The benefit isn't limited to stackalloc, but is proportional to how much state is on the stack that needs to be cleared.  As such, it generally is much more impactful with stackalloc, but it can show up as a peanut butter cost elsewhere.

```C#
[SkipLocalsInit]
[Benchmark]
public unsafe int WithSkipLocal()
{
    Guid g1, g2, g3, g4, g5, g6;

    // Do something cheap that let's the benchmark compile and stuff not get optimized away
    Unsafe.SkipInit(out g1);
    Unsafe.SkipInit(out g2);
    Unsafe.SkipInit(out g3);
    Unsafe.SkipInit(out g4);
    Unsafe.SkipInit(out g5);
    Unsafe.SkipInit(out g6);
    return *(int*)&g1 + *(int*)&g2 + *(int*)&g3 + *(int*)&g4 + *(int*)&g5 + *(int*)&g6;
}

[Benchmark]
public unsafe int WithoutSkipLocal()
{
    Guid g1, g2, g3, g4, g5, g6;

    // Do something cheap that let's the benchmark compile and stuff not get optimized away
    Unsafe.SkipInit(out g1);
    Unsafe.SkipInit(out g2);
    Unsafe.SkipInit(out g3);
    Unsafe.SkipInit(out g4);
    Unsafe.SkipInit(out g5);
    Unsafe.SkipInit(out g6);
    return *(int*)&g1 + *(int*)&g2 + *(int*)&g3 + *(int*)&g4 + *(int*)&g5 + *(int*)&g6;
}
```

|           Method |      Mean |     Error |    StdDev | Code Size |
|----------------- |----------:|----------:|----------:|----------:|
|    WithSkipLocal | 0.2983 ns | 0.0324 ns | 0.0303 ns |      33 B |
| WithoutSkipLocal | 0.7273 ns | 0.0015 ns | 0.0013 ns |      79 B |

If the goal was biggest bang for the buck, though, it's definitely where larger stackallocs are being employed.

Reply:
Another place it can show up impactfully is with fixed-size buffers, since as with stackallocs, they can often represent large regions of stack space that needs to be cleared, e.g.
```C#
[SkipLocalsInit]
[Benchmark]
public unsafe int WithSkipLocal()
{
    Data d;
    Unsafe.SkipInit(out d);
    return d.Ptr[0];
}

[Benchmark]
public unsafe int WithoutSkipLocal()
{
    Data d;
    Unsafe.SkipInit(out d);
    return d.Ptr[0];
}

unsafe struct Data
{
    public fixed byte Ptr[1024];
}
```

|           Method |       Mean |     Error |    StdDev | Code Size |
|----------------- |-----------:|----------:|----------:|----------:|
|    WithSkipLocal |  0.5009 ns | 0.0047 ns | 0.0041 ns |      64 B |
| WithoutSkipLocal | 14.8256 ns | 0.0411 ns | 0.0343 ns |     132 B |

Reply:
Given it comes with a degree of risk, it seems better not to add it unless here's a demonstrated benefit. I also assume you'd be targeting the hot paths like per-request processing. Hosting doesn't have much of those.

Reply:
I agree with @Tratcher *but* we also don't clear the array from the array pool so I don't see how this is different. Still I'd prefer doing this to some key placed in Kestrel rather than the entire assembly. That'll build more confidence.